compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp

```diff
@@ -380,6 +386,191 @@ void transform_dialect::ForallToWorkgroupOp::getEffects(
+  }
+
+  int64_t tripCount = 1;
+  for (auto [lb, ub, step] : llvm::zip_equal(lbs, ubs, steps)) {
+    tripCount *= (ub - lb) / step;
+  }
+  return tripCount;
+}
+
+LogicalResult compareWorkerCountsAndTypes(scf::ForallOp producer,
+                                          scf::ForallOp consumer) {
+  FailureOr<int64_t> producerTripCount = getTripCount(producer);
+  FailureOr<int64_t> consumerTripCount = getTripCount(consumer);
+  if (failed(producerTripCount) || failed(consumerTripCount) ||
+      *producerTripCount != *consumerTripCount) {
+    return failure();
+  }
+
+  mlir::TypeID mappingId = producer.getMapping()
+                               ->getValue()
+                               .front()
+                               .getAbstractAttribute()
+                               .getTypeID();
+  auto compareMappingTypes = [&](ArrayAttr array) {
+    return llvm::all_of(array.getValue(), [&](const Attribute &attr) {
+      return attr.getAbstractAttribute().getTypeID() == mappingId;
+    });
+  };
+
+  if (!compareMappingTypes(producer.getMappingAttr()) ||
+      !compareMappingTypes(consumer.getMappingAttr())) {
+    return failure();
+  }
+  return success();
+}
+
+Value getReplacementSlice(RewriterBase &rewriter, Location loc,
+                          tensor::ParallelInsertSliceOp parallelInsert,
+                          tensor::ExtractSliceOp extractSlice) {
+  return rewriter.create<IREE::GPU::ShuffleTensorOp>(
+      loc, extractSlice.getType(), parallelInsert.getSource(),
+      parallelInsert.getOffsets(), parallelInsert.getSizes(),
+      parallelInsert.getStrides(), parallelInsert.getStaticOffsets(),
+      parallelInsert.getStaticSizes(), parallelInsert.getStaticStrides(),
+      parallelInsert.getDest(), extractSlice.getOffsets(),
+      extractSlice.getSizes(), extractSlice.getStrides(),
+      extractSlice.getStaticOffsets(), extractSlice.getStaticSizes(),
+      extractSlice.getStaticStrides());
+}
+
+LogicalResult fuseForallIntoSlice(RewriterBase &rewriter,
+                                  scf::ForallOp producer,
+                                  scf::ForallOp consumer,
+                                  tensor::ExtractSliceOp slice) {
+  if (producer->getNumResults() != 1) {
+    return failure();
+  }
+
+  if (failed(compareWorkerCountsAndTypes(producer, consumer))) {
+    return failure();
+  }
+
+  auto isAll = [](ArrayRef<int64_t> array, int64_t cmp) {
+    return llvm::all_of(array, [cmp](int64_t x) { return x == cmp; });
+  };
+
+  if (!isAll(producer.getStaticStep(), 1) ||
+      !isAll(producer.getStaticLowerBound(), 0) ||
+      !isAll(consumer.getStaticStep(), 1) ||
+      !isAll(consumer.getStaticLowerBound(), 0)) {
+    return failure();
+  }
+
+  rewriter.setInsertionPoint(slice);
+
+  // Step 1. Compute the producer IDs in terms of the consumer IDs.
+
+  MLIRContext *context = rewriter.getContext();
+  Location loc = producer.getLoc();
+
+  AffineExpr linearId = getAffineConstantExpr(0, context);
```

Comment:
Make this simpler. Use an an AffineExpr which `s0 * s1`, start with a `OpFoldResult` of 1 and just apply the map repeatedly with compose affine apply. I know it creates a lot of maps, but its easier code to read.